### DISCUSSION TITLE
How to set up OpenAI using Google Gemini API

### DISCUSSION THREAD

**Discussion by y8s** [mealie-recipes/mealie#4541] [Q&A]
Is this even possible?
https://ai.google.dev/gemini-api/docs/openai

For example if I plop this in to my compose.yml it fails miserably. 
```
      - OPENAI_BASE_URL=https://generativelanguage.googleapis.com/v1beta/
      - OPENAI_API_KEY=FakeAPIKey_$DLKFsfJSd;lkfjSDfuLKSJDfKASJDGFUAQWYf*EWF
      - OPENAI_MODEL=gemini-1.5-flash
```
`mealie-omni-1  | openai.BadRequestError: Error code: 400 - [{'error': {'code': 400, 'message': 'Invalid JSON payload received. Unknown name "response_format": Cannot find field.', 'status': 'INVALID_ARGUMENT', 'details': [{'@type': 'type.googleapis.com/google.rpc.BadRequest', 'fieldViolations': [{'description': 'Invalid JSON payload received. Unknown name "response_format": Cannot find field.'}]}]}}]`

(Also for a second I thought the "upload an image" was an image of FOOD which would have been amazing if AI could create a recipe from a picture of the food dish.)

Thanks!

**Comment by michael-genson** [comment]
Seems like they don't support the full API spec yet (which makes sense since it's still in beta).
FYI: https://platform.openai.com/docs/guides/structured-outputs#function-calling-vs-response-format

We haven't switched to the new structured outputs yet (it didn't exist when we implemented the integration).

**Reply by y8s** [reply]
Ah ok. I was just looking at free alternatives to OpenAI. I also run oobabooga but not regularly enough to warrant pointing to it.

**Comment by KiahConnor** [comment]
OPENAI_BASE_URL=https://generativelanguage.googleapis.com/v1beta/openai/
OPENAI_API_KEY=Key generated here https://ai.google.dev/gemini-api/docs/api-key
OPENAI_MODEL=gemini-1.5-flash

This seemed to work intermittently (which is an improvement!), but mainly you just get "error code 500: The model is overloaded. Please try again later". Might have more success with a different model or paid access.

**Reply by bsthilaire** [reply]
Tried gemini today, so far, worked for 4 out of 4. The error might also be related to the time of day.

**Reply by y8s** [reply]
it works now, weird. glad I asked and tried again.

**Reply by KiahConnor** [reply]
Yeah okay, I think I just had a bad time of day. Did 40 or so images, including handwritten without issue. Glad the above settings helped.

**Reply by y8s** [reply]
the openai suffix was key!

but I really want the reverse image recipe gen!

**Reply by Borriborri** [reply]
so do I understand correctly that you guys are using google gemini and not the openai parser? since this is free?

**Reply by y8s** [reply]
@Borriborri yes, using gemini for free.

**Reply by nickmaleao** [reply]
I was able to use gemini to import a recipe from an image, but the smart ingredient parsing is failing. 

 File "/app/mealie/services/parser_services/openai/parser.py", line 95, in _parse
    raise Exception("Failed to parse OpenAI response") from e
Exception: Failed to parse OpenAI response

But if I try to use the debug tool it works, it's only failing in the menu when editing a recipe and trying to parse the ingredients using the openai option.

**Reply by nickmaleao** [reply]
I opened the issue [4829](https://github.com/mealie-recipes/mealie/issues/4829#issuecomment-2585717651) and it was closed attributing the problem to the gemini ai. Meanwhile Google has launched a new model [Gemini 2.0 Flash ](https://ai.google.dev/gemini-api/docs/models/gemini#gemini-2.0-flash) and with that model the ingredient parsing inside the recipe menu is working.

The [gemini free tier](https://ai.google.dev/pricing#1_5flash) has some limits, but here aren't prices available yet for that specific model, in this [reddit post](https://www.reddit.com/r/Bard/comments/1hnvnsd/comment/m53128e/) a user mentioned that the experimental models aren't chargeable, but I still have statistics usage and quotas in the api usage reports. I have hundreds of recipes to parse and will monitor the api usage.

![image](https://github.com/user-attachments/assets/f1da1640-8ff0-4da5-aea7-92c7df251d4f)

**Comment by pentacore** [comment]
Just wanted to Chime in, the new Gemini 2.0 model seems to be working nicely now.
```env
OPENAI_BASE_URL=https://generativelanguage.googleapis.com/v1beta
OPENAI_MODEL=gemini-2.0-flash
OPENAI_API_KEY=YOUR KEY
```

Havent had any issues as of yet, and the free tier is quite generous. For a casual user you're unlikely to hit its limits

**Comment by robuck86** [comment]
I tried gemini-2.0-flash on the free tier. Sometimes it works, but most times I get 'too many requests' errors. I don't use the parser that frequently, so I don't think I should be exceeding any limits. Anyone know why? 

`HTTP Request: POST https://generativelanguage.googleapis.com/v1beta/chat/completions "HTTP/1.1 429 Too Many Requests"
raise Exception(f"OpenAI Request Failed. {e.__class__.__name__}: {e}") from e
Exception: OpenAI Request Failed. RateLimitError: Error code: 429 - [{'error': {'code': 429, 'message': 'Resource has been exhausted (e.g. check quota).', 'status': 'RESOURCE_EXHAUSTED'}}]
`

**Reply by pentacore** [reply]
That is strange, considering its supposed to have quite generous limits. But i'm observing similar behavior now, so something might've changed

**Reply by pentacore** [reply]
I'm seeing this issue being reported in multiple other opensource projects as well since the past week. So i think this is something on Googles side

**Comment by dissidence** [comment]
are these setting going to be exposed in settings anytime soon, or am i just not looking in the correct spot?

**Reply by michael-genson** [reply]
They're environment variables: https://docs.mealie.io/documentation/getting-started/installation/backend-config/#openai

**Reply by dissidence** [reply]
i found that, i was just hoping to find them in the browser.  and not have to go through command line.
